airbyte-integrations/connectors/source-twilio-taskrouter/source_twilio_taskrouter/twilio_taskrouter.yaml

```diff
@@ -0,0 +1,36 @@
+version: "0.1.0"
+
+definitions:
+  selector:
+    extractor:
+      field_pointer: []
+  requester:
+    url_base: "https://taskrouter.twilio.com"
+    http_method: "GET"
+    authenticator:
+      type: BasicHttpAuthenticator
+      username: "{{ config['account_sid'] }}"
+      password: "{{ config['auth_token'] }}"
+  retriever:
+    record_selector:
+      $ref: "*ref(definitions.selector)"
+    paginator:
+      type: NoPagination
```

Comment:
I understand that only one workspace is fetched, so there's no need to paginate. But if `sid` is offered, you can fetch other resources under the workspace. Can you implement one that requires a paginator? 

And the [Twilio's paginator](https://www.twilio.com/docs/usage/twilios-response) is applied across all APIs, including taskrouter. Can you leverage codes in the existing [source-twilio](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-twilio/README.md)?

Reply:
Do you want me to create another workspace stream and have the user enter the resource name? Because certain resources have indicated to avoid use of Page query in the taskrouter api

Reply:
A paginator would be needed when the response is too long, or the resource set is huge. Consequently, the requester has to include the page number or another form of cursor to get the remaining parts. 

For workspaces, the API "will return the first 50 Workspaces. Supply a PageSize parameter to fetch more than 50 Workspaces." Maybe you can't create multiple workspaces, but I guess the resource set is huge on other streams, for example, [activities](https://www.twilio.com/docs/taskrouter/api/activity#read-multiple-activity-resources)

Reply:
Sure that works, so for the workspaces stream I'll have to add another optional parameter for the Page Size. Can you point me to a cdk doc on how to implement it? It would be great

Reply:
> "will return the first 50 Workspaces. Supply a PageSize parameter to fetch more than 50 Workspaces."

It comes from [Twilio's API doc](https://www.twilio.com/docs/taskrouter/api/workspace#list-all-workspaces), you may supply the parameter to get more than 50. And you may pick or implement a paginator in order to get other pages. For example, the PageSize is 100, and there are 500 entities, there will be 5 pages in total. Airbyte needs to send 5 requests to complete a full import. 

My suggestion is to have a `workspaces` stream that fetches all workspaces, and add another stream that depends on the **required** field workspaceId.

Reply:
This is the doc for [the paginator](https://docs.airbyte.com/connector-development/config-based/understanding-the-yaml-file/pagination)

Reply:
I'm having some trouble implementing the paginator, so I've set it up so that the user can enter a page_size input and page_size reads it as page_size: "{{config['number']}}" but how do i convert it into an integer from a string

Reply:
number is not part of spec.yaml. When you add it to the spec, you could set the type of an attribute to integer. 

My understanding of page_size is it defines the number of entities in a page. You could set a constant number, for example, 50. Even a customer is allowed to use a large number, say 65535, the server-side on twilio may cap it to another number, say 200. And if there are thousands of entities, you need to turn the next page by including the page number. 

https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-greenhouse/source_greenhouse/greenhouse.yaml 

source-greenhouse is a good example, the greenhouse includes the page number in the headers of a response.   It's an industry standard called [RFC 8288](https://www.rfc-editor.org/rfc/rfc8288). Twilio envelopes the entities, and set page_size, page_number and even the link to next page, called next_page_uri in the response body. Something like the following:

```
{
  "page": 0,
  "page_size": 50,
  "uri": "\/2010-04-01\/Accounts\/ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\/Calls.json",
  "first_page_uri": "\/2010-04-01\/Accounts\/ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\/Calls.json?Page=0&PageSize=50",
  "previous_page_uri": null,
  "next_page_uri": "\/2010-04-01\/Accounts\/ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\/Calls.json?Page=1&PageSize=50&AfterSid=CA228399228abecca920de212121",
  "calls": [
```

To test your pagination strategy, you don't need to create excessive resources. You could simply set page_size to a small number. For example, you have 6 workers, and the page size is 2, you need to paginate 3 times. After you run the read command, airbyte will tell you how many workers are read.

Reply:
source-greenhouse is a good example, the greenhouse includes the page number in the headers of a response. It's an industry standard called [RFC 8288](https://www.rfc-editor.org/rfc/rfc8288). Twilio envelopes the entities, and set page_size, page_number and even the link to next page, called next_page_uri in the response body. Something like the following:

Yes, I was confused at first when the response header did not indicate the page_size and number but as you explained it enveloping it makes sense. I'll add the paginator source to the workers stream with the example values you provided and see how it goes.

Reply:
okay I've made the changes let me know if there are any issues

Reply:
@YiyangLi could you take a look at the changes and see if I need anything else?